torch/csrc/autograd/forward_grad.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+
+namespace torch { namespace autograd {
+
+struct ForwardGrad;
+
+// This file contains two classes that are used to store forward AD gradients and
+// ensure that they are scoped properly.
+// Because forward AD runs concurently with the evaluation of the function, we need
+// a mechanism to separate different forward AD invocations and be able to compute the
+// right gradients. We model such invocations as levels here.
+// The particular scoping issue mentionned above has two main drive:
+//   - Ensure that we can conveniently use forward AD within a high level API without
+//     leaking the forward AD states outside.
+//   - Ensure that we can keep the level that we expose to the user API simple (an integer
+//     that represents the nesting depth) while avoiding confusions when the level index
+//     is re-used.
+
+// The important external APIs from this file are:
+//   - ForwardADLevel::get_next_idx() that can be used to enter a new level and get its index
+//   - ForwardADLevel::release_idx() that can be used to exit a given level.
+//   - ForwardGrad() can be used to store a given forward gradient that will handle the level
+//     tracking automatically.
+
+// The behavior above is achieved by ensuring that the ForwardGrad containing gradients for a
+// given level register themselves properly with the corresponding level.
+// On the other hand, the level, when it is released, will reset all the gradients for this
+// level on all the ForwardGrad.
+
+struct TORCH_API ForwardADLevel {
+    ForwardADLevel(uint64_t idx): idx_(idx) {}
+    ~ForwardADLevel();
+
+    static uint64_t get_next_idx();
+    static void release_idx(uint64_t idx);
+    static std::shared_ptr<ForwardADLevel> get_by_idx(uint64_t idx);
+
+    void erase(const std::shared_ptr<ForwardGrad>& grad) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        grads_.erase(grad);
+    }
+
+    void insert(const std::shared_ptr<ForwardGrad>& grad) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        grads_.insert(grad);
+    }
+
+private:
+    std::unordered_set<std::shared_ptr<ForwardGrad>> grads_;
```

Comment:
Doesn't this leak all the tangents until the end of the forward AD region?

Reply:
Apparently AutogradMeta has finalizers and does the clearing

Reply:
Because you are storing forward gradients on the tensors, I am thinking that you can use weak pointers. Essentially, the strong reference to tensors is via ForwardGrad (dual tensor has strong reference to tangent tensor), and then the level struct has a weak reference to all the tangent tensor containers (this set here). Then, if the tangent tensor becomes dead because the dual tensor goes dead, it gets released; everyone who isn't released automatically this way gets cleared out by the destructor of the level.